container/lighttpd: Add image build option

Ferruck opened the pull request:
The image includes files needed by the host systemd in order to be able
to automatically "attach" lighttpd to the host as a "portable service".

In order to use the host paths enforced by the host systemd, changes to
the lighttpd.conf were necessary, as well.

These changes are WIP since I haven't been able to test them with a
recent enough systemd. The mounted images looks good, though.

This needs BobBuildTool/basement#5

Ferruck committed:
container/lighttpd: Add image build option

The image includes files needed by the host systemd in order to be able
to automatically "attach" lighttpd to the host as a "portable service".

In order to use the host paths enforced by the host systemd, changes to
the lighttpd.conf were necessary, as well.

jkloetzke commented on recipes/containers/lighttpd.yaml:
```diff
@@ -12,32 +12,58 @@ depends:
 buildScript: |
-    # start from scratch
-    rm -rf *
+    install_common() {
```
No need to wrap it into a function if this is anyway invoked by all `multiPackage` entries.

Ferruck replied:
Well, actually I didn't know that... 😬

jkloetzke commented on recipes/containers/lighttpd.yaml:
```diff
@@ -12,32 +12,58 @@ depends:
+
+        packageScript: |
+            dd if=/dev/zero of=lighttpd.raw bs=1M count=64
+            mkfs.ext4 -d $1 lighttpd.raw
```
Why not something like squashfs? This will either waste space or will break if the package size grows.

Ferruck replied:
Never used squashfs, I'll have a look.

jkloetzke commented:
LGTM...

Ferruck commented:
I'll see if I'm able to spin up a minimal Arch/Fedora VM in order to test this.

jkloetzke committed:
basement: bump to 53cc5496

Require Bob 0.16rc too.

jkloetzke committed:
containers/lighttpd: add default index.html

Show something if the user forgot to mount something useful to /srv/www.

jkloetzke commented:
I've got the service running on Fedora 31. Well, kind of...

I have to run the service with the "trusted" profile. Otherwise it fails with:

```
lighttpd[4808]: /usr/sbin/lighttpd: error while loading shared libraries: libcrypto.so.1.0.0: cannot enable executable stack as shared object requires: Operation not permitted
```

This happens even when I disabled SELinux. I had to do this anyway because I could not use `portablectl` otherwise. Configuring `firewalld` was another obstacle.

I'm not sure if the docket stuff is still working...

Ferruck commented on recipes/containers/lighttpd.yaml:
```diff
@@ -51,14 +51,27 @@ multiPackage:
-            touch etc/os-release
-            touch etc/resolve.conf
-            mkdir -p proc sys dev run tmp var/tmp
+            ln -s ../usr/lib/os-release etc/os-release
```
Any specific reason to put it here?

jkloetzke replied:
https://www.freedesktop.org/software/systemd/man/os-release.html

Ferruck replied:
OK, I was just wondering because the portable application overview only explicitly mentions `/etc/os-release`...

Ferruck commented on recipes/containers/lighttpd/index.html:
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="iso-8859-1"?>
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
```
Honestly? :astonished: :wink:

jkloetzke replied:
I've copied it from the lighttpd 404 page... :laughing:

Ferruck commented:
Well, that's a bit more elaborated than what I've come up with, thanks. Nice to hear that you got it working.

Regarding your error: May it be that it's picking up the lib from the host?

jkloetzke commented:
> Regarding your error: May it be that it's picking up the lib from the host?

No I don't think this is happening. It looks like the libraries that we build, `libcrypto.so.1.0.0` in particular, seem to be incompatible with the default systemd security settings. :thinking:

jkloetzke commented:
I think we're done except for some README updates. I had to update OpenSSL to make the "executable stack" error go away...

Ferruck commented:
Did you ensure that docker is still working?

jkloetzke commented:
No. I did not have the time. But I would expect it's still working...

Ferruck commented:
Please give me some minutes to at least provide a less ancient HTML page, that's awkward.

jkloetzke commented:
Sure. The README will also need some update to describe how the portable service is built and how it's used...

Ferruck committed:
containers/lighttpd: Provide a modern default file

Ferruck committed:
containers/lighttpd: Provide PID file

When started within docker the PID file has to be present in order to
run lighttpd.

Ferruck committed:
README: Describe portable service usage

Ferruck commented:
Done. Could you please check the `README.md` section about the portable service? You're the only one who's actually done that... 😄

jkloetzke commented:
> Done. Could you please check the `README.md` section about the portable service? You're the only one who's actually done that... smile

Looks good. I've corrected only a small typo and squashed the fixup commits. I would merge it unless you have something to add...

Ferruck commented:
For me this is done now. LGTM
